src/textual/widgets/_placeholder.py

```diff
@@ -1,63 +1,166 @@
 from __future__ import annotations
 
-from rich import box
-from rich.align import Align
-from rich.console import RenderableType
-from rich.panel import Panel
-from rich.pretty import Pretty
-import rich.repr
-from rich.style import Style
+from itertools import cycle
 
 from .. import events
-from ..reactive import Reactive
-from ..widget import Widget
+from ..app import ComposeResult
+from ..css._error_tools import friendly_list
+from ..reactive import Reactive, reactive
+from ..widgets import Static
+from .._typing import Literal
 
+PlaceholderVariant = Literal["default", "size", "text"]
+_VALID_PLACEHOLDER_VARIANTS_ORDERED: list[PlaceholderVariant] = [
+    "default",
+    "size",
+    "text",
+]
+_VALID_PLACEHOLDER_VARIANTS: set[PlaceholderVariant] = set(
+    _VALID_PLACEHOLDER_VARIANTS_ORDERED
+)
+_PLACEHOLDER_BACKGROUND_COLORS = [
+    "#881177",
+    "#aa3355",
+    "#cc6666",
+    "#ee9944",
+    "#eedd00",
+    "#99dd55",
+    "#44dd88",
+    "#22ccbb",
+    "#00bbcc",
+    "#0099cc",
+    "#3366bb",
+    "#663399",
+]
 
-@rich.repr.auto(angular=False)
-class Placeholder(Widget, can_focus=True):
 
-    has_focus: Reactive[bool] = Reactive(False)
-    mouse_over: Reactive[bool] = Reactive(False)
+class InvalidPlaceholderVariant(Exception):
+    pass
+
+
+class _PlaceholderLabel(Static):
+    pass
+
+
+class Placeholder(Static):
+    """A simple placeholder widget to use before you build your custom widgets.
+
+    This placeholder has a couple of variants that show different data.
+    Clicking the placeholder cycles through the available variants, but a placeholder
+    can also be initialised in a specific variant.
+
+    The variants available are:
+        default: shows an identifier label or the ID of the placeholder.
+        size: shows the size of the placeholder.
+        text: shows some Lorem Ipsum text on the placeholder.
+    """
+
+    DEFAULT_CSS = """
+    Placeholder {
+        align: center middle;
+        overflow-y: auto;
+    }
+
+    Placeholder.-text {
+        padding: 1;
+    }
+
+    Placeholder > _PlaceholderLabel {
+        content-align: center middle;
+    }
+    """
+    # Consecutive placeholders get assigned consecutive colors.
+    COLORS = cycle(_PLACEHOLDER_BACKGROUND_COLORS)
+
+    variant: Reactive[PlaceholderVariant] = reactive("default")
+
+    @classmethod
+    def reset_color_cycle(cls) -> None:
+        """Reset the placeholder background color cycle."""
+        cls.COLORS = cycle(_PLACEHOLDER_BACKGROUND_COLORS)
 
     def __init__(
-        # parent class constructor signature:
         self,
-        *children: Widget,
+        variant: PlaceholderVariant = "default",
+        *,
+        label: str | None = None,
```

Comment:
I think the label should be a positional arg. A placeholder without a label is of limited use, so we should require it.

Reply:
I don't understand. "We should require it" means it shouldn't have a default value of `None`, right?
It could be either positional or keyword.

When a placeholder doesn't get a label, it uses its ID (or the string "Placeholder" if there is no ID) so I think a placeholder without an explicit label _is_ useful. Especially when you have a bunch of placeholders and are playing around with them with live CSS editing.

So, do I make it positional or leave it as a keyword parameter?
And do I remove the default value of `None` so users have to pass in `None` explicitly, or do I remove the ability to receive `None` altogether?

Reply:
I think positional with no default. You wouldn’t need an id for a placeholder typically, because when you come to style it you would use a custom widget.

Reply:
Your push in this direction makes me think I misunderstood how placeholders will be used.

My premise is that (1) placeholders are used _before_ the custom widgets exist, so to create and style a layout that uses placeholders can only work with IDs.

The only other alternative (2) that comes to me is to have empty classes for future placeholders inherit from `Placeholder`, something like:

```py
class MyFutureWidget(Placeholder):
    ...
```

If (2) is the intended usage pattern, then I 100% agree with your push, I'll make the change, and also edit the docs to explain this is the intended usage pattern.

If (2) is not how we want this to be used/if (1) is what we think will happen, then I think a placeholder should require the least amount of things possible, and thus I would push for `label` having a default value of `None`.

Either way, probably makes sense to move `label` to be the first parameter.

Reply:
> A placeholder without a label is of limited use

I don't understand this because the old placeholders didn't require labels and they were very useful - almost every time I've used a placeholder in the past has been without a label. Almost every example of the old placeholder in the Textual repo was also without a label.

Reply:
Works you not want some kind of label? By that I mean some text that identifies what the placeholder will become?

My thinking is the first arg should be something like “options side bar” or “user message”

Maybe then the label should be positional but with a default of None, defaulting to id as it is now.

Reply:
Commit [9cd592b](https://github.com/Textualize/textual/pull/1229/commits/9cd592bd044b2175bfb54358f3bdcebff2dc6d14) reorders the parameters of the `Placeholder` so that the first parameter is `label=None`.